src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -0,0 +1,406 @@
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+
+'''
+This module contains the InlineTrans transformation.
+
+'''
+
+from psyclone.errors import InternalError
+from psyclone.psyGen import Transformation
+from psyclone.psyir.nodes import (
+    Call, Routine, Reference, CodeBlock, Return, Literal)
+from psyclone.psyir.symbols import (ContainerSymbol, DataSymbol, ScalarType,
+                                    ImportInterface)
+from psyclone.psyir.transformations.transformation_error import (
+    TransformationError)
+
+
+class InlineTrans(Transformation):
+    '''
+    This transformation takes a Call and replaces it with the body of the
+    target routine. It is used as follows:
+
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.nodes import Call, Routine
+    >>> from psyclone.psyir.transformations import InlineTrans
+    >>> code = ("""
+    ... module test_mod
+    ... contains
+    ...   subroutine run_it()
+    ...   integer :: i
+    ...   real :: a(10)
+    ...   do i=1,10
+    ...     a(i) = 1.0
+    ...     call sub(a(i))
+    ...   end do
+    ...   end subroutine run_it
+    ...   subroutine sub(x)
+    ...     real, intent(inout) :: x
+    ...     x = 2.0*x
+    ...   end subroutine sub
+    ... end module test_mod"""
+    >>> psyir = FortranReader().psyir_from_source(code)
+    >>> call = psyir.walk(Call)[0]
+    >>> inline_trans = InlineTrans()
+    >>> inline_trans.apply(call)
+    >>> # Uncomment the following line to see a text view of the schedule
+    >>> # print(psyir.walk(Routine)[0].view())
+
+    The routine to be inlined must not contain any un-resolved symbols or
+    access any symbols that are defined in the parent container. Currently,
+    (#924) the routine must also be in the same source file as the call.
+
+    '''
+    def apply(self, node, options=None):
+        '''
+        Takes the body of the routine that is the target of the supplied
+        call and replaces the call with it.
+
+        :param node: target PSyIR node.
+        :type node: subclass of :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str, str]]
```

Comment:
It doesn't matter much in this transformation but I think in general options should be Optional[Dict[str, Any]] as sometimes the value is a boolean or a int

Reply:
Oh yes, fixed.